File: scripts/NESAnalyzer.py
# -*- coding: utf-8 -*-
# NESAnalyzer.py -- Ghidra headless post-analysis script for NES PRG-ROM (Jython)
#
# Runs after Ghidra's auto-analysis. Exports:
#   - $NES_DISASM_OUT/functions.json
#   - $NES_DISASM_OUT/register_accesses.json
#   - $NES_DISASM_OUT/bank_NN.asm
#   - $NES_DISASM_OUT/call_graph.json
#
# Environment variables:
#   NES_RESET       - RESET vector address (hex, no 0x prefix)
#   NES_NMI         - NMI vector address
#   NES_IRQ         - IRQ vector address
#   NES_DISASM_OUT  - output directory path
#   NES_BANK_IDX    - bank index for output naming

# @category NES
# @menupath Analysis.NES Analyzer

import os
import json

from ghidra.program.flatapi import FlatProgramAPI
from java.io import File


HW_REG_NAMES = {
    0x2000: "PPUCTRL",    0x2001: "PPUMASK",    0x2002: "PPUSTATUS",
    0x2003: "OAMADDR",    0x2004: "OAMDATA",    0x2005: "PPUSCROLL",
    0x2006: "PPUADDR",    0x2007: "PPUDATA",
    0x4000: "SQ1_VOL",    0x4001: "SQ1_SWEEP",  0x4002: "SQ1_LO",    0x4003: "SQ1_HI",
    0x4004: "SQ2_VOL",    0x4005: "SQ2_SWEEP",  0x4006: "SQ2_LO",    0x4007: "SQ2_HI",
    0x4008: "TRI_LINEAR", 0x400A: "TRI_LO",     0x400B: "TRI_HI",
    0x400C: "NOISE_VOL",  0x400E: "NOISE_LO",   0x400F: "NOISE_HI",
    0x4010: "DMC_FREQ",   0x4011: "DMC_RAW",    0x4012: "DMC_START", 0x4013: "DMC_LEN",
    0x4014: "OAMDMA",     0x4015: "APU_STATUS",  0x4016: "JOY1",     0x4017: "FRAME_CNT",
}


def run():
    out_dir = os.environ.get("NES_DISASM_OUT")
    reset_hex = os.environ.get("NES_RESET", "")
    nmi_hex = os.environ.get("NES_NMI", "")
    irq_hex = os.environ.get("NES_IRQ", "")
    bank_idx_str = os.environ.get("NES_BANK_IDX", "0")

    if not out_dir:
        println("ERROR: NES_DISASM_OUT env var not set")
        return

    out_path = File(out_dir)
    if not out_path.exists():
        out_path.mkdirs()

    api = FlatProgramAPI(currentProgram, monitor)
    af = currentProgram.getAddressFactory()
    listing = currentProgram.getListing()
    memory = currentProgram.getMemory()

    # Mark interrupt vectors as entry points and create functions
    vec_hexes = [reset_hex, nmi_hex, irq_hex]
    vec_names = ["RESET_HANDLER", "NMI_HANDLER", "IRQ_HANDLER"]
    for i in range(len(vec_hexes)):
        h = vec_hexes[i]
        if h and h.strip():
            try:
                addr_val = int(h, 16)
                a = af.getAddress("0x%04X" % addr_val)
                if a is not None:
                    api.createFunction(a, vec_names[i])
                    println("Created entry: %s @ %s" % (vec_names[i], str(a)))
            except Exception as e:
                println("WARNING: Could not parse vector address: %s (%s)" % (h, str(e)))

    # Run standard analysis
    analyzeAll(currentProgram)

    # Collect all functions
    functions = {}
    func_iter = listing.getFunctions(True)
    while func_iter.hasNext():
        func = func_iter.next()
        addr_str = "0x" + str(func.getEntryPoint()).upper()
        callers = []
        callees = []

        for caller in func.getCallingFunctions(monitor):
            callers.append("0x" + str(caller.getEntryPoint()).upper())
        for callee in func.getCalledFunctions(monitor):
            callees.append("0x" + str(callee.getEntryPoint()).upper())

        # Collect source ASM for this function
        src_asm = []
        body = func.getBody()
        func_instrs = listing.getInstructions(body, True)
        while func_instrs.hasNext():
            fi = func_instrs.next()
            src_asm.append("$%04X:  %-6s %s" % (
                fi.getAddress().getOffset(),
                fi.getMnemonicString(),
                fi.getDefaultOperandRepresentation(0)
            ))

        info = {
            "name": func.getName(),
            "start": addr_str,
            "end": "0x" + str(func.getBody().getMaxAddress()).upper(),
            "callers": callers,
            "callees": callees,
            "source_asm": "\n".join(src_asm),
        }
        functions[addr_str] = info

    # Collect hardware register accesses
    reg_accesses = []
    instr_iter = listing.getInstructions(True)
    while instr_iter.hasNext():
        instr = instr_iter.next()
        mnem = instr.getMnemonicString().upper()
        is_write = mnem in ("STA", "STX", "STY")
        is_read = mnem in ("LDA", "LDX", "LDY", "BIT")

        if not is_write and not is_read:
            continue

        for op in range(instr.getNumOperands()):
            ref = instr.getAddress(op)
            if ref is None:
                continue
            ref_addr = ref.getOffset()
            if ref_addr in HW_REG_NAMES:
                access = {
                    "address": "0x%s" % str(instr.getAddress()).upper(),
                    "hw_address": "0x%04X" % ref_addr,
                    "hw_name": HW_REG_NAMES[ref_addr],
                    "type": "PPU" if ref_addr < 0x4000 else "APU",
                    "access": "write" if is_write else "read",
                    "mnemonic": mnem,
                }
                reg_accesses.append(access)

    # Export disassembly for this single bank
    nes_bank_idx = int(bank_idx_str) if bank_idx_str else 0

    # Get the loaded memory range — find the block containing our code,
    # not the zero-page register block Ghidra creates for 6502
    blocks = memory.getBlocks()
    code_block = None
    for blk in blocks:
        if blk.getStart().getOffset() >= 0x8000:
            code_block = blk
            break
    if code_block is None:
        # Fallback to largest block
        largest_size = 0
        for blk in blocks:
            if blk.getSize() > largest_size:
                largest_size = blk.getSize()
                code_block = blk
    if code_block is None:
        println("ERROR: No suitable memory block found")
        return
    bank_start = code_block.getStart().getOffset()
    bank_end = code_block.getEnd().getOffset() + 1
    println("Memory block: %s $%04X-$%04X (%d bytes)" % (
        code_block.getName(), bank_start, bank_end - 1, bank_end - bank_start))

    sb = ["; NES PRG Bank %02d -- $%04X-$%04X\n" % (nes_bank_idx, bank_start, bank_end - 1)]

    start_a = af.getAddress("0x%04X" % bank_start)
    if start_a is not None:
        # Get all function entry addresses
        func_addrs = set()
        all_funcs = listing.getFunctions(True)
        while all_funcs.hasNext():
            func_addrs.add(all_funcs.next().getEntryPoint().getOffset())

        bank_iter = listing.getInstructions(start_a, True)
        while bank_iter.hasNext():
            instr = bank_iter.next()
            instr_addr = instr.getAddress().getOffset()
            if instr_addr >= bank_end:
                break
            # Emit function label if this is a function entry
            if instr_addr in func_addrs:
                f = listing.getFunctionAt(instr.getAddress())
                if f is not None:
                    sb.append("%s:" % f.getName())
            sb.append("    %-6s %s" % (
                instr.getMnemonicString(),
                instr.getDefaultOperandRepresentation(0)
            ))

    bank_path = os.path.join(out_dir, "bank_%02d.asm" % nes_bank_idx)
    with open(bank_path, "w") as fh:
        fh.write("\n".join(sb))

    # Write JSON outputs
    funcs_path = os.path.join(out_dir, "functions.json")
    with open(funcs_path, "w") as fh:
        json.dump(functions, fh, indent=2)

    regs_path = os.path.join(out_dir, "register_accesses.json")
    with open(regs_path, "w") as fh:
        json.dump(reg_accesses, fh, indent=2)

    # call_graph.json: flat list of edges
    edges = []
    for addr_str, info in functions.items():
        for to_addr in info.get("callees", []):
            edges.append({"from": addr_str, "to": to_addr})

    cg_path = os.path.join(out_dir, "call_graph.json")
    with open(cg_path, "w") as fh:
        json.dump(edges, fh, indent=2)

    println("NESAnalyzer: bank %d -- exported %d functions, %d register accesses" % (
        nes_bank_idx, len(functions), len(reg_accesses)))


try:
    run()
except Exception as e:
    import traceback
    println("FATAL ERROR in NESAnalyzer.py: " + str(e))
    println(traceback.format_exc())

File: scripts/NESAnalyzer.java
// NESAnalyzer.java — Ghidra headless post-analysis script for NES PRG-ROM
//
// Runs after Ghidra's auto-analysis. Exports:
//   - $NES_DISASM_OUT/functions.json       — function boundaries + call graph
//   - $NES_DISASM_OUT/register_accesses.json — all NES hardware register access sites
//   - $NES_DISASM_OUT/bank_NN.asm          — per-bank disassembly listings
//   - $NES_DISASM_OUT/call_graph.json      — caller/callee pairs
//
// Environment variables consumed:
//   NES_RESET       — RESET vector address (hex, no 0x prefix)
//   NES_NMI         — NMI vector address
//   NES_IRQ         — IRQ vector address
//   NES_DISASM_OUT  — output directory path

//@category NES
//@menupath Analysis.NES Analyzer

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.*;
import ghidra.program.model.listing.*;
import ghidra.program.model.mem.*;
import ghidra.program.model.pcode.*;
import ghidra.program.model.symbol.*;
import ghidra.program.flatapi.FlatProgramAPI;

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class NESAnalyzer extends GhidraScript {

    // NES hardware register ranges
    private static final long PPU_BASE  = 0x2000L;
    private static final long PPU_END   = 0x4000L;
    private static final long APU_BASE  = 0x4000L;
    private static final long APU_END   = 0x4018L;

    private static final Map<Long, String> HW_REG_NAMES = new LinkedHashMap<>();
    static {
        HW_REG_NAMES.put(0x2000L, "PPUCTRL");
        HW_REG_NAMES.put(0x2001L, "PPUMASK");
        HW_REG_NAMES.put(0x2002L, "PPUSTATUS");
        HW_REG_NAMES.put(0x2003L, "OAMADDR");
        HW_REG_NAMES.put(0x2004L, "OAMDATA");
        HW_REG_NAMES.put(0x2005L, "PPUSCROLL");
        HW_REG_NAMES.put(0x2006L, "PPUADDR");
        HW_REG_NAMES.put(0x2007L, "PPUDATA");
        HW_REG_NAMES.put(0x4000L, "SQ1_VOL");
        HW_REG_NAMES.put(0x4001L, "SQ1_SWEEP");
        HW_REG_NAMES.put(0x4002L, "SQ1_LO");
        HW_REG_NAMES.put(0x4003L, "SQ1_HI");
        HW_REG_NAMES.put(0x4004L, "SQ2_VOL");
        HW_REG_NAMES.put(0x4005L, "SQ2_SWEEP");
        HW_REG_NAMES.put(0x4006L, "SQ2_LO");
        HW_REG_NAMES.put(0x4007L, "SQ2_HI");
        HW_REG_NAMES.put(0x4008L, "TRI_LINEAR");
        HW_REG_NAMES.put(0x400AL, "TRI_LO");
        HW_REG_NAMES.put(0x400BL, "TRI_HI");
        HW_REG_NAMES.put(0x400CL, "NOISE_VOL");
        HW_REG_NAMES.put(0x400EL, "NOISE_LO");
        HW_REG_NAMES.put(0x400FL, "NOISE_HI");
        HW_REG_NAMES.put(0x4010L, "DMC_FREQ");
        HW_REG_NAMES.put(0x4011L, "DMC_RAW");
        HW_REG_NAMES.put(0x4012L, "DMC_START");
        HW_REG_NAMES.put(0x4013L, "DMC_LEN");
        HW_REG_NAMES.put(0x4014L, "OAMDMA");
        HW_REG_NAMES.put(0x4015L, "APU_STATUS");
        HW_REG_NAMES.put(0x4016L, "JOY1");
        HW_REG_NAMES.put(0x4017L, "FRAME_CNT");
    }

    @Override
    public void run() throws Exception {
        String outDir     = System.getenv("NES_DISASM_OUT");
        String resetHex   = System.getenv("NES_RESET");
        String nmiHex     = System.getenv("NES_NMI");
        String irqHex     = System.getenv("NES_IRQ");

        if (outDir == null) {
            println("ERROR: NES_DISASM_OUT env var not set");
            return;
        }

        Files.createDirectories(Paths.get(outDir));

        FlatProgramAPI api = new FlatProgramAPI(currentProgram, monitor);
        AddressFactory af  = currentProgram.getAddressFactory();
        Listing listing    = currentProgram.getListing();
        Memory memory      = currentProgram.getMemory();

        // Mark interrupt vectors as entry points and create functions
        String[] vecHexes  = {resetHex, nmiHex, irqHex};
        String[] vecNames  = {"RESET_HANDLER", "NMI_HANDLER", "IRQ_HANDLER"};
        for (int i = 0; i < vecHexes.length; i++) {
            if (vecHexes[i] != null && !vecHexes[i].isEmpty()) {
                try {
                    long addr = Long.parseLong(vecHexes[i], 16);
                    Address a = af.getAddress(String.format("0x%04X", addr));
                    if (a != null) {
                        api.createFunction(a, vecNames[i]);
                        println("Created entry: " + vecNames[i] + " @ " + a);
                    }
                } catch (NumberFormatException e) {
                    println("WARNING: Could not parse vector address: " + vecHexes[i]);
                }
            }
        }

        // Run standard analysis
        analyzeAll(currentProgram);

        // Collect all functions
        Map<String, Map<String, Object>> functions = new LinkedHashMap<>();
        FunctionIterator funcIter = listing.getFunctions(true);
        while (funcIter.hasNext()) {
            Function func = funcIter.next();
            String addrStr = "0x" + func.getEntryPoint().toString().toUpperCase();
            List<String> callers = new ArrayList<>();
            List<String> callees = new ArrayList<>();

            for (Function caller : func.getCallingFunctions(monitor)) {
                callers.add("0x" + caller.getEntryPoint().toString().toUpperCase());
            }
            for (Function callee : func.getCalledFunctions(monitor)) {
                callees.add("0x" + callee.getEntryPoint().toString().toUpperCase());
            }

            Map<String, Object> info = new LinkedHashMap<>();
            info.put("name",    func.getName());
            info.put("start",   addrStr);
            info.put("end",     "0x" + func.getBody().getMaxAddress().toString().toUpperCase());
            info.put("callers", callers);
            info.put("callees", callees);
            functions.put(addrStr, info);
        }

        // Collect hardware register accesses
        List<Map<String, String>> regAccesses = new ArrayList<>();
        InstructionIterator instrIter = listing.getInstructions(true);
        while (instrIter.hasNext()) {
            Instruction instr = instrIter.next();
            String mnem = instr.getMnemonicString().toUpperCase();
            boolean isWrite = mnem.equals("STA") || mnem.equals("STX") || mnem.equals("STY");
            boolean isRead  = mnem.equals("LDA") || mnem.equals("LDX") || mnem.equals("LDY") || mnem.equals("BIT");

            if (!isWrite && !isRead) continue;

            for (int op = 0; op < instr.getNumOperands(); op++) {
                Address ref = instr.getAddress(op);
                if (ref == null) continue;
                long refAddr = ref.getOffset();
                if (HW_REG_NAMES.containsKey(refAddr)) {
                    Map<String, String> access = new LinkedHashMap<>();
                    access.put("address",    "0x" + instr.getAddress().toString().toUpperCase());
                    access.put("hw_address", String.format("0x%04X", refAddr));
                    access.put("hw_name",    HW_REG_NAMES.get(refAddr));
                    access.put("type",       refAddr < 0x4000 ? "PPU" : "APU");
                    access.put("access",     isWrite ? "write" : "read");
                    access.put("mnemonic",   mnem);
                    regAccesses.add(access);
                }
            }
        }

        // Export per-bank disassembly listings
        long baseAddr    = 0x8000L;
        int  bankSz16kb  = 0x4000;
        MemoryBlock[] blocks = memory.getBlocks();
        // Find the total PRG size from the memory block
        long totalSize = 0;
        for (MemoryBlock b : blocks) {
            totalSize = Math.max(totalSize, b.getEnd().getOffset() - baseAddr + 1);
        }
        int numBanks = (int) Math.max(1, (totalSize + bankSz16kb - 1) / bankSz16kb);

        for (int bankIdx = 0; bankIdx < numBanks; bankIdx++) {
            long bankStart = baseAddr + (long) bankIdx * bankSz16kb;
            long bankEnd   = bankStart + bankSz16kb;
            StringBuilder sb = new StringBuilder();
            sb.append(String.format("; NES PRG Bank %02d — $%04X–$%04X\n", bankIdx, bankStart, bankEnd - 1));
            sb.append(String.format(".org $%04X\n\n", bankStart));

            Address startA = af.getAddress(String.format("0x%04X", bankStart));
            Address endA   = af.getAddress(String.format("0x%04X", bankEnd - 1));
            if (startA == null) continue;

            InstructionIterator bankIter = listing.getInstructions(startA, true);
            while (bankIter.hasNext()) {
                Instruction instr = bankIter.next();
                if (instr.getAddress().getOffset() >= bankEnd) break;
                sb.append(String.format("$%04X:  %-6s %s\n",
                    instr.getAddress().getOffset(),
                    instr.getMnemonicString(),
                    instr.getDefaultOperandRepresentation(0)));
            }

            Path bankPath = Paths.get(outDir, String.format("bank_%02d.asm", bankIdx));
            Files.writeString(bankPath, sb.toString());
        }

        // Write JSON outputs
        writeJson(Paths.get(outDir, "functions.json"), toJson(functions));
        writeJson(Paths.get(outDir, "register_accesses.json"), toJson(regAccesses));

        // call_graph.json: flat list of edges
        List<Map<String, String>> edges = new ArrayList<>();
        for (Map.Entry<String, Map<String, Object>> e : functions.entrySet()) {
            String from = e.getKey();
            List<String> callees = (List<String>) e.getValue().get("callees");
            for (String to : callees) {
                Map<String, String> edge = new LinkedHashMap<>();
                edge.put("from", from);
                edge.put("to", to);
                edges.add(edge);
            }
        }
        writeJson(Paths.get(outDir, "call_graph.json"), toJson(edges));

        println(String.format("NESAnalyzer: exported %d functions, %d register accesses, %d banks",
            functions.size(), regAccesses.size(), numBanks));
    }

    private static void writeJson(Path path, String json) throws IOException {
        Files.writeString(path, json);
    }

    // Minimal JSON serializer (avoids external deps in Ghidra context)
    private static String toJson(Object obj) {
        if (obj instanceof Map) {
            StringBuilder sb = new StringBuilder("{\n");
            Map<?, ?> map = (Map<?, ?>) obj;
            int i = 0;
            for (Map.Entry<?, ?> e : map.entrySet()) {
                sb.append("  ").append(jsonStr(e.getKey().toString()))
                  .append(": ").append(toJson(e.getValue()));
                if (++i < map.size()) sb.append(",");
                sb.append("\n");
            }
            return sb.append("}").toString();
        } else if (obj instanceof List) {
            StringBuilder sb = new StringBuilder("[\n");
            List<?> list = (List<?>) obj;
            for (int i = 0; i < list.size(); i++) {
                sb.append("  ").append(toJson(list.get(i)));
                if (i < list.size() - 1) sb.append(",");
                sb.append("\n");
            }
            return sb.append("]").toString();
        } else if (obj instanceof Boolean) {
            return obj.toString();
        } else if (obj instanceof Number) {
            return obj.toString();
        } else if (obj == null) {
            return "null";
        } else {
            return jsonStr(obj.toString());
        }
    }

    private static String jsonStr(String s) {
        return "\"" + s.replace("\\", "\\\\").replace("\"", "\\\"")
                       .replace("\n", "\\n").replace("\r", "") + "\"";
    }
}
